File: athanore/plugins/builtin/overview.py
"""The overview pane: what a run cost, what it is, and what it did.

09 §Builtins are plugins gives it one row — "run pane (`dashboard`:
metric tiles, token bars, `kv` meta, `table` nodes)" — and 10 §Panes
fixes every field it shows. One route and one panel: the panel is a
``dashboard`` in the ``run`` slot, and the route is where its data comes
from.

The response is the ``dashboard`` shape of 09 §Panel kinds — ``note``,
``metrics``, ``table`` — plus a ``meta`` object, which is the two-column
`kv` grid 10 puts between the tiles and the NODES table. A renderer that
knows only the ``dashboard`` kind draws the first three and ignores the
fourth; the SPA's own overview renderer draws all of it. Adding a key is
what the declared kind allows; inventing a kind for one pane is not.

**Nothing is zero-filled.** A metric whose fact is unknown is *absent*
(01 §Real data only), so a run on which no agent has yet reported
anything shows two tiles rather than four claiming zero tokens and no
cost. The same rule runs through the node table — a node with no stats
entry carries no ``tokens`` key — and through ``meta``, where SESSION is
there only once an agent has run.

10's per-node token bars are drawn from the same ``table`` rows the NODES
grid is: the bar and the cell are one number, and sending it twice is how
they would come to differ.

**The totals are the run's totals.** They come from ``RunRepo.detail``,
the one read ``GET /api/runs/{id}`` makes, so the tile and
``RunDetail.stats`` are the same number from the same query rather than
two sums that can drift.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from athanore.plugins.context import PluginContext
from athanore.store.clock import now
from athanore.store.rows import RunRow, RunStats, TaskRow, TaskStatus
from athanore.workflow import Workflow

__all__ = ["declare"]

#: The columns of the NODES table, in the order 10 §Panes draws them.
#: ``kind`` is the column hint of 09 §Panel kinds: it says what the value
#: is, not how wide to draw it.
COLUMNS: list[dict[str, str]] = [
    {"key": "node", "label": "NODE", "kind": "text"},
    {"key": "attempts", "label": "ATT", "kind": "number"},
    {"key": "status", "label": "STATUS", "kind": "status"},
    {"key": "tokens", "label": "TOKENS", "kind": "number"},
    {"key": "duration_s", "label": "DUR", "kind": "duration"},
]

#: The attempt statuses that put a run "in" a node right now. The same
#: pair ``RunRepo.list`` aggregates ``current_nodes`` from and the same
#: pair ``GET /api/runs/{id}`` recomputes, so the pane and the run row
#: cannot disagree about which nodes are live.
IN_FLIGHT = frozenset({TaskStatus.in_progress, TaskStatus.waiting})


def declare(wf_host: Workflow) -> None:
    """Declare the overview route and its pane on ``wf_host``."""

    @wf_host.route("/overview")
    async def overview(ctx: PluginContext) -> dict[str, Any]:
        """A run's totals, its meta, and a row per node it has entered."""

        run, tasks, stats = await ctx.services.run.detail()
        runs = [row.id for row in await ctx.services.run.list()]
        body: dict[str, Any] = {
            "metrics": _metrics(run, stats, runs),
            "table": {"columns": COLUMNS, "rows": _nodes(tasks)},
            "meta": _meta(run, tasks),
        }
        if run.description:
            body["note"] = run.description
        return body

    wf_host.panel(
        "overview",
        slot="run",
        kind="dashboard",
        source=overview,
        refresh_on=["run.*", "task.*", "agent.stats"],
    )


def _metrics(run: RunRow, stats: RunStats, runs: list[str]) -> list[dict[str, Any]]:
    """The four tiles of 10 §Panes, minus the ones nothing is known for.

    TOKENS and COST are the run's totals over **every** attempt, failed
    ones included — they were paid for (05 §Stats entry) — and each is
    absent when no attempt reported it. DURATION is wall-clock,
    ``finished − created`` or ``now − created``, which is a fact about
    every run from the moment it exists. POSITION is the run's place in
    the dispatch list, and it is the one tile whose value is a string:
    "4 of 12" is two numbers read together, and a tile shows one value.
    """

    metrics: list[dict[str, Any]] = []
    if stats.total_tokens is not None:
        metrics.append({"label": "TOKENS", "value": stats.total_tokens})
    if stats.cost is not None:
        metrics.append({"label": "COST", "value": stats.cost})
    metrics.append({"label": "DURATION", "value": _elapsed(run)})
    position = _position(run, runs)
    if position is not None:
        metrics.append({"label": "POSITION", "value": position})
    return metrics


def _elapsed(run: RunRow) -> float:
    """How long the run has been going, in seconds (10 §Panes)."""

    end = run.finished if run.finished is not None else now()
    return (end - run.created).total_seconds()


def _position(run: RunRow, runs: list[str]) -> str | None:
    """``"4 of 12"``: where the run sits in the dispatch list.

    ``runs`` is the list in dispatch order, which for the builtin scope
    is every run there is. ``None`` when the run is not in it — a run
    deleted between the two reads — because a place in a list the run is
    not in is not a fact (01 §Real data only).
    """

    if run.id not in runs:
        return None
    return f"{runs.index(run.id) + 1} of {len(runs)}"


def _meta(run: RunRow, tasks: list[TaskRow]) -> dict[str, Any]:
    """The `kv` grid of 10 §Panes, in its order.

    STATUS carries the nodes the run is in right now, joined the way the
    run list joins them. SESSION is the agent session of the most recent
    attempt that reported one, whole rather than shortened — how many
    characters of it to show is the renderer's decision. AGENTS counts
    the attempts that carry a stats entry. Both are absent until an agent
    has run, and DESCRIPTION until the operator has written one — which
    is not the same absence: 10 §Panes drops SESSION and AGENTS from the
    grid and draws "DESCRIPTION is the run's, or `—`", so the renderer
    supplies the dash for the key this omits (15, D161).
    """

    meta: dict[str, Any] = {
        "RUN": run.id,
        "WORKFLOW": run.workflow,
        "TITLE": run.title,
        "STATUS": _status(run, tasks),
        "AGE": (now() - run.created).total_seconds(),
    }
    session = _session(tasks)
    if session is not None:
        meta["SESSION"] = session
    agents = sum(1 for task in tasks if task.stats)
    if agents:
        meta["AGENTS"] = agents
    if run.description:
        meta["DESCRIPTION"] = run.description
    return meta


def _status(run: RunRow, tasks: list[TaskRow]) -> str:
    """The run status, and the nodes it is in when it is in any.

    10 §Panes asks for the nodes on a ``running`` run; the condition here
    is that there *are* in-flight attempts, which is the same set on a
    running run and is honest on the ones where the two disagree — a
    paused run whose last attempt is still finishing has a node, and
    hiding it would be hiding the reason it has not stopped yet.
    """

    live = sorted({task.node for task in tasks if task.status in IN_FLIGHT})
    if not live:
        return run.status.value
    return f"{run.status.value} · {' · '.join(live)}"


def _session(tasks: list[TaskRow]) -> str | None:
    """The ``session_id`` of the most recent attempt that reported one."""

    for task in reversed(tasks):
        session = (task.stats or {}).get("session_id")
        if isinstance(session, str) and session:
            return session
    return None


def _nodes(tasks: list[TaskRow]) -> list[dict[str, Any]]:
    """One row per node the run has entered, in the order it entered them.

    Attempts of one node collapse into one row: ATT is how many there
    have been, STATUS is the latest one's, TOKENS is what they spent
    between them, and DUR is how long they took. A node the run has not
    reached has no attempt and therefore no row — the graph pane is
    where the shape of the whole workflow is drawn, and this table is the
    history.

    DUR is wall-clock, ``finished − started`` summed over the attempts
    that have both, to match the DURATION tile; the agent's own
    ``duration_s`` is a different measurement and lives in the stats
    entry. A row omits ``tokens`` or ``duration_s`` outright when nothing
    is known, rather than showing a zero nobody measured.
    """

    rows: dict[str, dict[str, Any]] = {}
    for task in tasks:
        row = rows.setdefault(
            task.node, {"node": task.node, "attempts": 0, "status": task.status.value}
        )
        row["attempts"] += 1
        row["status"] = task.status.value
        tokens = (task.stats or {}).get("total_tokens")
        if isinstance(tokens, int):
            row["tokens"] = row.get("tokens", 0) + tokens
        elapsed = _attempt_seconds(task)
        if elapsed is not None:
            row["duration_s"] = row.get("duration_s", 0.0) + elapsed
    return list(rows.values())


def _attempt_seconds(task: TaskRow) -> float | None:
    """How long one attempt took, or ``None`` if it is not known yet."""

    started: datetime | None = task.started
    finished: datetime | None = task.finished
    if started is None or finished is None:
        return None
    return (finished - started).total_seconds()

File: web/src/panes/kinds/overview.ts
/**
 * How the overview pane reads what its source sent
 * (`docs/v1/10-frontend.md` §Panes — "Field sources").
 *
 * Separate from `./Overview.tsx` for the reason `./format.ts` is
 * separate from the renderers: a component module that also exports
 * functions is one React Fast Refresh cannot update in place
 * (`.oxlintrc.json`, `react/only-export-components`).
 *
 * Everything here is the same rule twice over. **A value is formatted by
 * the field it is, not by its type**: `612,884` is a token count and
 * `$0.2914` is money, and the generic `formatValue` of `./format.ts`
 * would print both as bare numbers because it does not know which is
 * which. And **nothing is invented**: a field the source omitted stays
 * omitted, a bar is drawn only for a node that reported a number, and a
 * span nobody measured reads as `—` rather than as a zero (01 §Real data
 * only).
 */
import { humaniseElapsed } from '../../components/RunList'
import type { RunOutput, TaskView } from '../../api/gen/types.gen'
import { formatValue } from './format'
import type { Metric, TableRow } from './shape'

/** How many characters of the agent session the SESSION field shows. */
export const SESSION_LENGTH = 8

/** The narrowest a token bar is drawn, so a small one is still visible. */
export const MIN_BAR_PERCENT = 3

/** How much of a terminal value the OUTPUTS list previews. */
export const PREVIEW_LENGTH = 120

/** What a field with nothing behind it reads as, as everywhere else. */
const DASH = '—'

/** The two meta fields the grid is the wrong shape for (10 §Panes). */
const TITLE_KEY = 'TITLE'
const DESCRIPTION_KEY = 'DESCRIPTION'

/**
 * `612,884` — a count, grouped.
 *
 * `en-US` rather than the browser's locale: the mock's separators are
 * the app's, and a page that grouped with spaces in one country and
 * commas in another would not be the design.
 */
const GROUPED = new Intl.NumberFormat('en-US')

/** A count, or `—` when the number is not one. */
export function formatCount(value: unknown): string {
  return typeof value === 'number' && Number.isFinite(value)
    ? GROUPED.format(value)
    : formatValue(value)
}

/**
 * `$0.2914` — money, to the mock's four decimal places.
 *
 * Agent costs run to fractions of a cent per attempt, so two places
 * would round most of a run's tiles to `$0.00`.
 */
export function formatCost(value: unknown): string {
  return typeof value === 'number' && Number.isFinite(value)
    ? `$${value.toFixed(4)}`
    : formatValue(value)
}

/**
 * `246s` — a span in seconds, as the mock writes both the DURATION tile
 * and the NODES table's DUR column (15, D161).
 *
 * One decimal below ten seconds, because a node that took 0.4 s did not
 * take none; whole seconds above it, because the tenth of a second in
 * `3944.2s` is noise.
 */
export function formatSeconds(value: unknown): string {
  if (typeof value !== 'number' || !Number.isFinite(value)) return formatValue(value)
  const seconds = Math.max(0, value)
  return seconds < 10 ? `${String(Number(seconds.toFixed(1)))}s` : `${String(Math.round(seconds))}s`
}

/**
 * One of the four tiles, formatted by its label (10 §Panes).
 *
 * A label this build does not know — a metric a later Athanore adds — is
 * printed by the generic formatter rather than dropped: an unknown field
 * is still the server's data.
 */
export function formatMetric(metric: Metric): Metric {
  switch (metric.label) {
    case 'TOKENS':
      return { label: metric.label, value: formatCount(metric.value) }
    case 'COST':
      return { label: metric.label, value: formatCost(metric.value) }
    case 'DURATION':
      return { label: metric.label, value: formatSeconds(metric.value) }
    default:
      return { label: metric.label, value: formatValue(metric.value) }
  }
}

/**
 * The `kv` meta grid, formatted field by field (10 §Panes).
 *
 * The source sends AGE as seconds and SESSION whole; 10 asks for the
 * mock's humanised age and for eight characters of the session, and
 * both of those are the renderer's decision rather than the route's.
 * Key order is the source's own, which is the order 10 lists the fields
 * in, so a field this build has no rule for still lands in its place.
 *
 * **TITLE and DESCRIPTION are deliberately not here** — a function
 * called `formatMeta` that drops two of the meta keys is a thing to put
 * back by accident. They are {@link formatAbout}'s, because the grid
 * they used to sit in is `minmax(240px, 1fr)` wide and a description is
 * written in a textarea and can be a paragraph (15, D208); the block
 * that draws them has the pane's whole width.
 */
export function formatMeta(meta: Record<string, unknown>): Record<string, string> {
  const out: Record<string, string> = {}
  for (const [key, value] of Object.entries(meta)) {
    if (key === TITLE_KEY || key === DESCRIPTION_KEY) continue

    if (key === 'AGE') {
      out[key] =
        typeof value === 'number' && Number.isFinite(value)
          ? humaniseElapsed(value * 1000)
          : formatValue(value)
    } else if (key === 'SESSION') {
      out[key] =
        typeof value === 'string' ? value.slice(0, SESSION_LENGTH) : formatValue(value)
    } else if (key === 'AGENTS') {
      out[key] = formatCount(value)
    } else {
      out[key] = formatValue(value)
    }
  }
  return out
}

/** The two fields the grid is the wrong shape for (10 §Panes). */
export type RunAbout = {
  /** The run's title, or `null` when the source sent none. */
  title: string | null
  /** The run's description, or `—` (10 §Panes, D161). */
  description: string
}

/**
 * The full-width TITLE and DESCRIPTION block, or nothing at all.
 *
 * DESCRIPTION is the one field drawn whether or not the source sent it:
 * 10 §Panes says it "is the run's, or `—`", against the SESSION and
 * AGENTS beside it, which it says are omitted until an agent has run.
 * TITLE is not dashed — 10 gives the dash to DESCRIPTION and to nothing
 * else, and the builtin's route always sends TITLE, so an absent one
 * came from a panel that meant not to have one, and its row is dropped.
 *
 * `null` when the source sent **neither** key: an `overview`-kind panel
 * a plugin declares need not be about a run at all (09 §Panel kinds),
 * and drawing `TITLE —` and `DESCRIPTION —` under one would be this
 * renderer inventing two fields nobody sent (01 §Real data only).
 *
 * Both values go through `formatValue` and nothing else — no preview,
 * no truncation, no re-wrapping. The block is a layout, not a formatter.
 */
export function formatAbout(meta: Record<string, unknown>): RunAbout | null {
  const hasTitle = TITLE_KEY in meta
  if (!hasTitle && !(DESCRIPTION_KEY in meta)) return null
  return {
    title: hasTitle ? formatValue(meta[TITLE_KEY]) : null,
    description: DESCRIPTION_KEY in meta ? formatValue(meta[DESCRIPTION_KEY]) : DASH,
  }
}

/** One per-node token bar: what it says, how long it is, and its colour. */
export type TokenBar = {
  node: string
  /** The count itself, grouped — the bar's right-hand column. */
  count: string
  /** Its length as a percentage of the largest node's, `1`–`100`. */
  percent: number
  /** Whether the run is in this node right now (accent, not accent-700). */
  active: boolean
}

/**
 * The bars, from the NODES rows the same table draws (10 §Panes — "token
 * bars scale to the largest node total in the run").
 *
 * **A node with no `tokens` key gets no bar.** The route omits the key
 * for a node no agent reported on (`overview.py`), and a zero-length bar
 * would be this renderer claiming it spent nothing. A node that really
 * reported `0` keeps its bar, at the minimum width.
 */
export function tokenBars(
  rows: readonly TableRow[],
  activeNodes: readonly string[],
): TokenBar[] {
  const counted = rows.filter(
    (row) => typeof row['tokens'] === 'number' && Number.isFinite(row['tokens']),
  )
  const totals = counted.map((row) => row['tokens'] as number)
  const largest = Math.max(0, ...totals)

  return counted.map((row, index) => {
    const total = totals[index] ?? 0
    const scaled = largest > 0 ? Math.round((total / largest) * 100) : 0
    return {
      node: String(row['node'] ?? ''),
      count: GROUPED.format(total),
      percent: Math.max(MIN_BAR_PERCENT, scaled),
      active: activeNodes.includes(String(row['node'] ?? '')),
    }
  })
}

/**
 * The attempt a NODES row opens in the task drawer: the node's most
 * recent one.
 *
 * The row collapses every attempt of its node into one line whose STATUS
 * is the latest attempt's (`overview.py`), so the drawer the operator
 * expects from clicking it is that same attempt. `RunDetail.tasks` is
 * oldest first (08 §Runs), so the last match is it.
 */
export function latestTaskOf(
  tasks: readonly TaskView[] | undefined,
  node: string,
): number | undefined {
  let latest: number | undefined
  for (const task of tasks ?? []) if (task.node === node) latest = task.id
  return latest
}

/** One line of the OUTPUTS list: which branch ended, and with what. */
export type OutputLine = {
  /** Stable within the list: a terminal attempt appears once. */
  taskId: number
  /** The node, with its branch keys when the run fanned out. */
  node: string
  /** The value, printed and cut to {@link PREVIEW_LENGTH}. */
  preview: string
}

/** `qa [1: alpha]` — the node, and the branch it terminated in. */
function branchLabel(output: RunOutput): string {
  const branch = output.branch ?? []
  if (branch.length === 0) return output.node
  const keys = branch.map(
    (frame) =>
      `${String(frame.index)}${frame.key === undefined ? '' : `: ${formatValue(frame.key)}`}`,
  )
  return `${output.node} [${keys.join(' · ')}]`
}

/** A value as one line, cut where it stops being a preview. */
export function preview(value: unknown, limit: number = PREVIEW_LENGTH): string {
  const text = formatValue(value).replace(/\s+/g, ' ').trim()
  if (text === '') return DASH
  return text.length <= limit ? text : `${text.slice(0, limit)}…`
}

/**
 * The OUTPUTS list, or nothing at all.
 *
 * 10 §Panes shows it "when a run had more than one terminal branch", and
 * a run with one terminal task has its value in the run's own output
 * rather than in a list of one (04 §Routing edge cases). So one entry —
 * the linear case, and the fan-out closed by a join — draws no section.
 */
export function outputLines(outputs: readonly RunOutput[] | undefined): OutputLine[] {
  const list = outputs ?? []
  if (list.length < 2) return []
  return list.map((output) => ({
    taskId: output.task_id,
    node: branchLabel(output),
    preview: preview(output.value),
  }))
}
